### PR TITLE
Build binding with PyQt6

### PR DESCRIPTION
**Pull request by troopa81**
This PR is a first work to build the binding with PyQt6. So far, approximatively half of the python tests are successfull.

I choose not to update the sip files in order to bring as less as possible modifications to the actual PyQt5 binding, and so lower the chance of regression. As a consequence sipify_all is called before building.

### Enum issue

- When using Enum/Flag in PyQt6, you have to [specify scope](https://www.riverbankcomputing.com/static/Docs/PyQt5/gotchas.html#enums), I generated mock for QGIS enum so QGIS tests could work but all Qt enum would need to be rewritten... That's an issue that we wouldn't have if we choose to switch from PyQt to PySide ( [see](https://www.pythonguis.com/faq/pyqt6-vs-pyside6/) Namespaces & Enum)

- I mocked the Qt enums [here](https://github.com/troopa81/QGIS/blob/build_qt6_ci_pyqt/python/testing/__init__.py#L67) so the tests could work

- flags needs to be generated with the SIP instruction /BaseType=IntFlag/, if not there are considered as plain enum and some operator are missing ('&' operator for instance)
- Problem is that int flag make them unsigned int and so enum with negative values ([QgsProcessing::SourceType](https://github.com/qgis/QGIS/blob/master/src/core/processing/qgsprocessing.h#L46) for instance) get overflow value
- Even with int flag there is no implicit conversion from int to enum like it is expected in some test ( [here for instance](https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tools/dataobjects.py#L72) )

### To be done
  - Fix all the todo in conversions.sip
  - replace QTextCodec with QStringConverter in QgsVectorDataProvider and QgsJsonUtils
  - Fix all tests
  -  Fix all `TODO Make it work with Qt6` ( disabled with % If( !PYQT6 ) )

### CODE REVIEW

File: .github/workflows/run-tests.yml [hunk 1]
```diff
@@ -52,7 +52,7 @@ jobs:
             # LD_PRELOAD: /lib/x86_64-linux-gnu/libSegFault.so
             experimental: false
 
-          - distro-version: '36'
+          - distro-version: '38'
```
Comment (nyalldawson) [inline .github/workflows/run-tests.yml:55]: I suggest opening a seperate pr with just this change -- there's likely a bit of work required to get this update passing which would best be done as a seperate effort.

Comment (troopa81) [reply .github/workflows/run-tests.yml:55]: Normally, I already take care of some issues in #52638 regarding builing on fedora 38 and it should work out of the box. Still, I'll propose a separate PR. Unfortunately clang16 has landed in fedora 38 in the meantime and there is a version [incompatibility](https://bugzilla.redhat.com/show_bug.cgi?id=2187234) with clazy

**Comment by nyalldawson** [comment]
@troopa81 can you pull some more of the self-contained bits out of this? eg https://github.com/qgis/QGIS/pull/52739/commits/9a6146254031218d7e9ee1c6366d59891da012e9 and https://github.com/qgis/QGIS/pull/52739/commits/3609c25c6d8a866782cbbd66601661f6a1e4f8c3 could be merged easily and would help reduce the size and complexity of this pr

**Comment by troopa81** [comment]
I blocked all the failing tests. 

@nyalldawson Do we agree on the general approach of this PR?

I would like to have it merged so I can move on in fixing tests in separate PRs.

@elpaso @3nids @m-kuhn Any thoughts about this?

File: python/plugins/processing/gui/BatchInputSelectionPanel.py [hunk 2]
```diff
@@ -23,9 +23,16 @@
...
-from qgis.PyQt.QtWidgets import QWidget, QHBoxLayout, QMenu, QPushButton, QLineEdit, QSizePolicy, QAction, QFileDialog
+from qgis.PyQt.QtWidgets import QWidget, QHBoxLayout, QMenu, QPushButton, QLineEdit, QSizePolicy, QFileDialog
 from qgis.PyQt.QtGui import QCursor
 
+# TODO QGIS 4.0: Fix code if PyQt6
+from qgis.PyQt.QtCore import QT_VERSION
+if (QT_VERSION >= 0x060000):
+    from qgis.PyQt.QtGui import QAction
+else:
+    from qgis.PyQt.QtWidgets import QAction
```
Comment (m-kuhn) [inline python/plugins/processing/gui/BatchInputSelectionPanel.py:34]: Could you make QAction available from qgis.PyQt.QtGui for the PyQt5 version instead?
I believe I've done that for Qt4 as well a while back

Comment (troopa81) [reply python/plugins/processing/gui/BatchInputSelectionPanel.py]: Done [e4a7acc](https://github.com/qgis/QGIS/pull/52739/commits/e4a7acc8b7d4c579b93b55269b10c8ae25497972)

**Comment by m-kuhn** [comment]
Sorry for the super short review @troopa81 it's a quite large PR and I currently lack the time for an in-depth review.
Potentially someone from the people who are working on behalf of the PSC on the Qt6 migration could have a look at this?

**Pull request closed without merging**